File: 2022/day11.py
import queue
# TODO: replace this with collections.deque since you don't need threading stuff
# and queue.Queue.get blocks. I tried using queue.Queue on day12 and it hanged on
# the second BFS

prime_prod = 1

class Monkey:

    def _init_queue(self, line):
        self._q = queue.Queue()
        for x in line[16:].split(', '):
            self._q.put(int(x))

    def _init_op(self, line):
        op = line[21:]  # it's stripped!
        assert op[0] in '+*'
        if op == '+ old':
            self._op = lambda y : y + y
        elif op == '* old':
            self._op = lambda y : y * y
        else:
            op_val = int(op[2:])
            if op[0] == '+':
                self._op = lambda y : y + op_val
            else:
                self._op = lambda y : y * op_val

    def _init_test(self, line):
        self._test_val = int(line[19:])  # stripped

    def work(self, monkeys):
        while self._q.qsize() != 0:
            self._count += 1
            item = self._q.get()
            item = self._op(item)# % self._test_val
            if self._part == 2:
                item %= prime_prod
            if self._part == 1:
                item //= 3
            other = self._throw[0 == item % self._test_val]
            monkeys[other]._q.put(item)

    def __init__(self, lines, part):
        self._count = 0
        self._part = part
        self._init_queue(lines[1])
        self._init_test(lines[3])
        self._init_op(lines[2])
        self._throw = dict()
        self._throw[True] = int(lines[4][25:])
        self._throw[False] = int(lines[5][26:])


def prod_best_two(monkeys):
    best_two = sorted([m._count for m in monkeys], reverse=True)[:2]
    return best_two[0] * best_two[1]


with open('data.txt', 'r') as file:
    paragraphs = [p.strip() for p in file.read().split('\n\n')]
    monkeys1 = []
    monkeys2 = []
    for p in paragraphs:
        lines = [l.strip() for l in p.split('\n')]
        monkeys1.append(Monkey(lines, 1))
        monkeys2.append(Monkey(lines, 2))
    for i in range(20):
        for m in monkeys1:
            m.work(monkeys1)
    print(prod_best_two(monkeys1))
    # part 2
    # assumes all the _test_val are primes
    # not sure why this works, but it's something to do with
    # chinese remainder theorem
    for m in monkeys2:
        prime_prod *= m._test_val
    for i in range(10000):
        for m in monkeys2:
            m.work(monkeys2)
    print(prod_best_two(monkeys2))

File: 2022/day11.go
package main

import (
	"fmt"
	"math"
	"os"
	"strconv"
	"strings"
)

var primeProd int64 = 1

type monkey struct {
	q          queue[int64]
	op         func(int64) int64
	testVal    int
	count      int
	part       int
	trueOther  int
	falseOther int
}

func (m *monkey) work(monkeys *[]monkey) {
	for !m.q.isEmpty() {
		m.count++
		item := m.q.remove()
		item = m.op(item)
		switch m.part {
		case 1:
			item /= 3
		case 2:
			item %= primeProd
		}
		var other int
		if 0 == item%int64(m.testVal) {
			other = m.trueOther
		} else {
			other = m.falseOther
		}
		(*monkeys)[other].q.add(item)
	}
}

func add(x int64, y int64) int64 {
	return x + y
}

func mul(x int64, y int64) int64 {
	return x * y
}

func binaryOpFromStr(s string) func(int64, int64) int64 {
	switch s {
	case "+":
		return add
	case "*":
		return mul
	default:
		panic(s)
	}
}

func unaryOpFromString(s string) func(int64) int64 {
	operands := strings.Split(s, " ")
	binOp := binaryOpFromStr(operands[1])
	if operands[2] == "old" {
		return func(x int64) int64 {
			return binOp(x, x)
		}
	} else {
		c, _ := strconv.Atoi(operands[2])
		return func(x int64) int64 {
			return binOp(x, int64(c))
		}
	}
}

func fromBlock(block string) monkey {
	lines := strings.Split(block, "\n")
	items := strings.Split(strings.TrimPrefix(lines[1], "  Starting items: "), ", ")
	q := queue[int64]{nil, nil}
	for _, item := range items {
		x, _ := strconv.Atoi(item)
		q.add(int64(x))
	}
	opStr := strings.TrimPrefix(lines[2], "  Operation: new = ")
	op := unaryOpFromString(opStr)
	testStr := strings.TrimPrefix(lines[3], "  Test: divisible by ")
	testVal, _ := strconv.Atoi(testStr)
	trueStr := strings.TrimPrefix(lines[4], "    If true: throw to monkey ")
	trueOther, _ := strconv.Atoi(trueStr)
	falseStr := strings.TrimPrefix(lines[5], "    If false: throw to monkey ")
	falseOther, _ := strconv.Atoi(falseStr)
	return monkey{q, op, testVal, 0, 1, trueOther, falseOther}
}

func prodBestTwo(monkeys *[]monkey) int64 {
	best := math.MinInt
	second := math.MinInt
	for i := range *monkeys {
		c := (*monkeys)[i].count
		if c == best {
			second = c
		} else if c > best {
			second = best
			best = c
		} else if c > second {
			second = c
		}
	}
	return int64(best) * int64(second)
}

func main() {
	filename := "/home/xdavidliu/Documents/aoc/input11.txt"
	bytesRead, _ := os.ReadFile(filename)
	blocks := strings.Split(string(bytesRead), "\n\n")
	var monkeys1 []monkey = nil
	var monkeys2 []monkey = nil
	for _, block := range blocks {
		m1 := fromBlock(block)
		m2 := m1
		m2.q = m2.q.copy()
		m2.part = 2
		monkeys1 = append(monkeys1, m1)
		monkeys2 = append(monkeys2, m2)
	}
	for i := 0; i < 20; i++ {
		for k := range monkeys1 {
			monkeys1[k].work(&monkeys1)
		}
	}
	fmt.Println("part 1 =", prodBestTwo(&monkeys1)) // 107822
	for i := range monkeys2 {
		primeProd *= int64(monkeys2[i].testVal)
	}
	for i := 0; i < 10_000; i++ {
		for k := range monkeys2 {
			monkeys2[k].work(&monkeys2)
		}
	}
	fmt.Println("part 2 =", prodBestTwo(&monkeys2)) // 27267163742
}
